pkg/kubelet/kubelet.go

```diff
@@ -2867,6 +2867,7 @@ func (kl *Kubelet) syncNetworkStatus() {
 
 // Set addresses for the node.
 func (kl *Kubelet) setNodeAddress(node *api.Node) error {
+	hostnameAddress := api.NodeAddress{Type: api.NodeHostName, Address: kl.GetHostname()}
```

Comment:
@liggitt: Ok, trying to outline it:

```
$ git grep -A3 "func .* CurrentNodeName"
pkg/cloudprovider/providers/aws/aws.go:func (c *AWSCloud) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/aws/aws.go-	return c.selfAWSInstance.nodeName, nil
pkg/cloudprovider/providers/aws/aws.go-}
pkg/cloudprovider/providers/aws/aws.go-
--
pkg/cloudprovider/providers/fake/fake.go:func (f *FakeCloud) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/fake/fake.go-	return hostname, nil
pkg/cloudprovider/providers/fake/fake.go-}
pkg/cloudprovider/providers/fake/fake.go-
--
pkg/cloudprovider/providers/gce/gce.go:func (gce *GCECloud) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/gce/gce.go-	return hostname, nil
pkg/cloudprovider/providers/gce/gce.go-}
pkg/cloudprovider/providers/gce/gce.go-
--
pkg/cloudprovider/providers/mesos/mesos.go:func (c *MesosCloud) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/mesos/mesos.go-	return hostname, nil
pkg/cloudprovider/providers/mesos/mesos.go-}
pkg/cloudprovider/providers/mesos/mesos.go-
--
pkg/cloudprovider/providers/openstack/openstack.go:func (i *Instances) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/openstack/openstack.go-	return hostname, nil
pkg/cloudprovider/providers/openstack/openstack.go-}
pkg/cloudprovider/providers/openstack/openstack.go-
--
pkg/cloudprovider/providers/ovirt/ovirt.go:func (v *OVirtCloud) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/ovirt/ovirt.go-	return hostname, nil
pkg/cloudprovider/providers/ovirt/ovirt.go-}
pkg/cloudprovider/providers/ovirt/ovirt.go-
--
pkg/cloudprovider/providers/rackspace/rackspace.go:func (i *Instances) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/rackspace/rackspace.go-	// Beware when changing this, nodename == hostname assumption is crucial to
pkg/cloudprovider/providers/rackspace/rackspace.go-	// apiserver => kubelet communication.
pkg/cloudprovider/providers/rackspace/rackspace.go-	return hostname, nil
--
pkg/cloudprovider/providers/vsphere/vsphere.go:func (i *Instances) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/vsphere/vsphere.go-	return i.localInstanceID, nil
pkg/cloudprovider/providers/vsphere/vsphere.go-}
pkg/cloudprovider/providers/vsphere/vsphere.go-
```

Most map directly to hostname, VSphere and AWS don't. 

* VSphere maps to VM name, not necessarily resolvable. Right now the VM name has to be configured to be a master-resolvable DNS name.

https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/vsphere/vsphere.go#L127

* AWS gets its hostname from the API, using `instance.PrivateDnsName`, which is equal to an EC2 instance hostname:

  https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go#L982

The current kubelet code will get a Node's hostname by executing `uname -n`, unless `--hostname-override` is set, then it will use this value.

In general I don't think this is wrong (actually it's the intended behavior IMO): It's better to require the Node's hostname to be master-resolvable than have a convention that the Nodename has to be resolvable.

However this could introduce a breaking change for clusters set up under the assumption: Nodename matters, hostname doesn't. To make sure this doesn't happen we can add the `Type:Hostname` Address to the address array in the affected cloud-providers code already, and add it in the kubelet code only when it's not set by the cloud-provider (actually this sanity-check should be there anyway).